[Route Groups] Support AddFilter, WithOpenApi and other additive conventions

halter73 opened the pull request:
The nice thing about this fix is that more local "conventions" can see and modify metadata added by outer groups. At first, I wasn't sure this was going to be possible for groups, but I found the solution by adding the new virtual `EndpointDataSource.GetGroupedEndpoints(RouteGroupContext)` method, and then creating a custom `RouteEndointDataSource` that takes over a lot of the complicated logic that used to exist in `EndpointRouteBuilderExtensions`.

The upside of moving this logic into `RouteEndointDataSource` other than general cleanliness, is that as a custom `EndpointDataSource`, it can override `EndpointDataSource.GetGroupedEndpoints(GroupContext)` and inspect the full group prefix and all the group metadata before calling `RequestDelegateFactory.Create()` or running any filters.

Even though `RequestDelegateFactory` now runs after any conventions are added to the endpoint (aside from conventions added by the `RequestDelegateFactory`), `WithOpenApi` is fixed by having the `RouteEndointDataSource` add `MethodInfo` before running any conventions. `RouteEndointDataSource` also adds any attributes as metadata. This is more similar to the previous behavior of `RequestDelegateFactory` in .NET 6 where it did not add this metadata.

The new-to-.NET-7 `RequestDelegateFactoryOptions.InitialEndpointMetadata` is now just `RequestDelegateFactoryOptions.EndpointMetadata` because now it's mutable and is equivalent to the `RouteEndpointBuilder.Metadata` at this very late stage of building. This gives the absolute most flexibility to filters and metadata providers because they can add metadata wherever they want, but they have to be careful not to override metadata they don't want to by simply adding to the end without looking if anything has already been configured. I think this is fine because the filter and metadata provider APIs are new .NET 7 so there aren't expectations that metadata added will have a low precedence yet. @captainsafia @DamianEdwards 

You can see how this works by looking at some of the tests. For example, `WithOpenApi_GroupMetadataCanBeSeenByAndOverriddenByMoreLocalMetadata`:

```csharp
string GetString() => "Foo";

static void WithLocalSummary(RouteHandlerBuilder builder)
{
    builder.WithOpenApi(operation =>
    {
        operation.Summary += $" | Local Summary | 200 Status Response Content-Type: {operation.Responses["200"].Content.Keys.Single()}";
        return operation;
    });
}

WithLocalSummary(builder.MapDelete("/root", GetString));

var outerGroup = builder.MapGroup("/outer");
var innerGroup = outerGroup.MapGroup("/inner");

WithLocalSummary(outerGroup.MapDelete("/outer-a", GetString));

// The order WithOpenApi() is relative to the MapDelete() methods does not matter.
outerGroup.WithOpenApi(operation =>
{
    operation.Summary = "Outer Group Summary";
    return operation;
});

WithLocalSummary(outerGroup.MapDelete("/outer-b", GetString));
WithLocalSummary(innerGroup.MapDelete("/inner-a", GetString));

innerGroup.WithOpenApi(operation =>
{
    operation.Summary += " | Inner Group Summary";
    return operation;
});

WithLocalSummary(innerGroup.MapDelete("/inner-b", GetString));

var summaries = builder.DataSources.SelectMany(ds => ds.Endpoints).Select(e => e.Metadata.GetMetadata<OpenApiOperation>().Summary).ToArray();

Assert.Equal(5, summaries.Length);
Assert.Contains(" | Local Summary | 200 Status Response Content-Type: text/plain", summaries);
Assert.Contains("Outer Group Summary | Local Summary | 200 Status Response Content-Type: text/plain", summaries);
Assert.Contains("Outer Group Summary | Local Summary | 200 Status Response Content-Type: text/plain", summaries);
Assert.Contains("Outer Group Summary | Inner Group Summary | Local Summary | 200 Status Response Content-Type: text/plain", summaries);
Assert.Contains("Outer Group Summary | Inner Group Summary | Local Summary | 200 Status Response Content-Type: text/plain", summaries);
```

And `AddRouteHandlerFilterMethods_WorkWithMapGroup`.

```csharp
string PrintId(int id) => $"ID: {id}";

static void AddParamIncrementingFilter(IEndpointConventionBuilder builder)
{
    builder.AddRouteHandlerFilter(async (context, next) =>
    {
        context.Arguments[0] = ((int)context.Arguments[0]!) + 1;
        return await next(context);
    });
}

AddParamIncrementingFilter(builder.Map("/{id}", PrintId));

var outerGroup = builder.MapGroup("/outer");
AddParamIncrementingFilter(outerGroup);
AddParamIncrementingFilter(outerGroup.Map("/{id}", PrintId));

var innerGroup = outerGroup.MapGroup("/inner");
AddParamIncrementingFilter(innerGroup);
AddParamIncrementingFilter(innerGroup.Map("/{id}", PrintId));

var endpoints = builder.DataSources
    .SelectMany(ds => ds.Endpoints)
    .ToDictionary(e => ((RouteEndpoint)e).RoutePattern.RawText!);

Assert.Equal(3, endpoints.Count);

// For each layer of grouping, another filter is applies which increments the expectedId by 1 each time.
await AssertIdAsync(endpoints["/{id}"], expectedPattern: "/{id}", expectedId: 3);
await AssertIdAsync(endpoints["/outer/{id}"], expectedPattern: "/outer/{id}", expectedId: 4);
await AssertIdAsync(endpoints["/outer/inner/{id}"], expectedPattern: "/outer/inner/{id}", expectedId: 5);
```

Fixes #41427
Fixes #42137
Fixes #41722

halter73 committed:
Add EndpointDataSource.GetGroupedEndpoints

halter73 committed:
TBuilderify AddFilter and WithOpenApi

halter73 committed:
Fix missed ServiceProvider -> ApplicationServices

halter73 committed:
Fix some tests and doc comments

halter73 committed:
Fix CompositeEndpointDataSource.HandleChange ordering

- Always call user code outside of our lock.

captainsafia commented on src/Http/Http.Abstractions/src/RequestDelegateResult.cs:
```diff
@@ -23,7 +23,8 @@ public RequestDelegateResult(RequestDelegate requestDelegate, IReadOnlyList<obje
 
     /// <summary>
-    /// Gets endpoint metadata inferred from creating the <see cref="RequestDelegate" />
+    /// Gets endpoint metadata inferred from creating the <see cref="RequestDelegate" />. This does not contain
```
This comment is a little confusing only because `InitialEndpointMetadat` no longer exists, correct?

captainsafia commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -174,27 +174,31 @@ public static RequestDelegateResult Create(MethodInfo methodInfo, Func<HttpConte
+    }
 
-        if (options?.InitialEndpointMetadata is not null)
+    private static List<object> GetMetadataList(IList<object> metadata)
```
I'm a little confused by this method. Isn't `metadata` going to be a `List<object>` already if we're reading it from the factory context?

captainsafia replied:
Nevermind. I see the `FactoryContext.Metadata` type has been changed to an `IList`. Why not use the concrete type there?

halter73 replied:
It's because we're taking in a mutable `IList<object>` which has to match `EndpointBuilder.Metadata` and we need to return an `IReadOnlyList<object>` which is required by the already-shipped `RequestDelegateResult.RequestDelegateResult`. If it's just a `List<object>` like it is most of the time we can cast it and treat it as an `IReadOnlyList<object>`.

I also could have just tried `metadata is IReadOnlyList<object> readOnlyList` and returned that if it succeeded and only return `new List<object>(metadata)` as a last resort. Not sure why I didn't do it that way to begin with. Fixed. I'm not calling it `AsReadOnlyList` instead to make the intent more clear.

BrennanConroy commented on src/Http/Routing/test/UnitTests/Builder/GroupTest.cs:
```diff
@@ -301,7 +294,7 @@ public async Task ChangingMostEndpointBuilderPropertiesInConvention_Works()
         Assert.False(mapGetCalled);
         Assert.True(replacementCalled);
-        Assert.Equal("Replaced!", endpoint.DisplayName);
+        Assert.Equal("HTTP: GET Replaced!", endpoint.DisplayName);
```
Is this a preferred change?

halter73 replied:
Good eye! This is a natural consequence of group conventions now actually having a lower precedence than endpoint-specific conventions. In this case, we have the endpoint-specific convention being added by `MapMethods`.

https://github.com/dotnet/aspnetcore/blob/228ff77ba14d493d2765486f4f9726d50ad7dab3/src/Http/Routing/src/Builder/EndpointRouteBuilderExtensions.cs#L350-L351

So, I think this is the preferred behavior, and we just didn't notice that there was a group-specific convention that was getting clobbered before.

If we decided we wanted a group replacing the display name not to be wrapped by $"HTTP: {string.Join(", ", httpMethods)} {b.DisplayName}", it would actually be quite easy to do this. We would just move the wrapping logic straight into [the start of RouteEndpointDataSource.CreateRouteEndpointBuilder()](https://github.com/dotnet/aspnetcore/pull/42195/files#diff-b193e1f6a7930155befe00ee0d8d2e5c6e23951a1f5046c50e02d9c2527ac7e1R91-R100). The ease and flexibility of making that change is part of why I like adding the `RouteEndpointDataSource` so much.

halter73 replied:
After further consideration, I decided ASP.NET developers aren't going to appreciate the nuance of the display name getting modified by an "endpoint-specific convention" vs just having that name to begin with. Since it was easy to fix, I did so. You can see the changes in [this commit](https://github.com/dotnet/aspnetcore/pull/42195/commits/1c5feef7f549e7bbf8c5c44a8724de30815864c0) which I think clean things up even further. EndpointRouteBuilderExtensions is now very simple!

BrennanConroy commented on src/Http/Routing/src/Builder/RouteHandlerFilterExtensions.cs:
```diff
@@ -56,27 +57,77 @@ public static RouteHandlerBuilder AddFilter(this RouteHandlerBuilder builder, IR
     /// </summary>
     /// <param name="builder">The <see cref="RouteHandlerBuilder"/>.</param>
-    /// <param name="routeHandlerFilter">A <see cref="Delegate"/> representing the core logic of the filter.</param>
+    /// <param name="routeHandlerFilter">A <see cref="Func{T1, T2, TResult}"/> representing the core logic of the filter.</param>
```
Not a fan of either `Delegate` or `Func<T1, T2, TResult>`. Maybe just say "A method representing ..."

halter73 committed:
Address PR review feedback

halter73 committed:
Address more PR feedback

halter73 committed:
Move MapMethods logic into RouteEndpointDataSource

halter73 committed:
Merge remote-tracking branch 'origin/main' into halter73/41427

halter73 committed:
Fix internal doc comment cref

halter73 committed:
Move MapFallback logic to RouteEndpointDataSource too

halter73 committed:
GetGroupedEndpoints -> GetEndpointGroup

halter73 committed:
Fix build

- Fix nullability in EndpointMetadataApiDescriptionProviderTest
- Remove redundant ArgumentNullException checks

halter73 committed:
I thought I already fixed this lol

- 97a5ada5d7c6bb547eaa58203e052ecc29f99c76

captainsafia commented on src/OpenApi/src/OpenApiRouteHandlerBuilderExtensions.cs:
```diff
@@ -46,37 +36,72 @@ public static RouteHandlerBuilder WithOpenApi(this RouteHandlerBuilder builder)
+                if (configure is not null)
                 {
-                    routeEndpointBuilder.Metadata.Add(configureOperation(openApiOperation));
+                    var configuredOperation = configure(existingOperation);
```
If I'm understanding this correctly, the callback provided to `WithOpenApi` will run on operations inserted by the parent group. So if I have something like:

```
var outer = app.MapGroup("/1").WithOpenApi();
var inner = outer.MapGroup("/2").WithOpenApi();
var nested = inner.MapGroup("/3").WithOpenApi(o => o.Summary = "foo");
var anotherOne = inner.MapGroup("/4").WithOpenApi();
```

Then the operation in the summary will be applied to the operations in `inner` and `outer`? What would `anotherOne` see in this case?

halter73 replied:
Assuming that there was an endpoint added to `anotherOne` via `MapGet` or something, the path would be prefixed by "/1/2/4" (no 3 since `anotherOne` isn't under `nested`).

It would not see the "foo" summary since that was applied to the `nested`, but it would see anything done by callbacks to the `outer` and `inner` group. In this case, that's nothing.

Any endpoint defined under `nested` would have a "foo" summary unless overridden by a more nested group or on the endpoint specifically.

halter73 commented on src/Http/Routing/src/RouteGroupBuilder.cs:
```diff
@@ -10,131 +10,96 @@ namespace Microsoft.AspNetCore.Routing;
+
+            if (_routeGroupBuilder._dataSources.Count is 1)
+            {
+                return _routeGroupBuilder._dataSources[0].GetEndpointGroup(new RouteGroupContext(fullPrefix, combinedConventions, applicationServices));
```
Note from @davidfowl. Reuse the `RouteGroupContext`, since it's immutable.

captainsafia commented on src/Http/Routing/src/Builder/RouteHandlerFilterExtensions.cs:
```diff
@@ -56,27 +57,77 @@ public static RouteHandlerBuilder AddFilter(this RouteHandlerBuilder builder, IR
+        {
+            // Even if we split RouteFilterMetadata across multiple metadata objects, RouteEndointDataSource will still use all of them.
+            // We're just trying to be good citizens and keep the Endpoint.Metadata as small as possible.
+            RouteFilterMetadata? filterMetadata = null;
```
Nice! I like that this has the side-effect of making filters registered on an endpoint inspectable.

halter73 replied:
I agree that it's a lot easier to inspect with a debugger this way.

I did make `RouteFilterMetadata` internal for now, so it's not really inspectable programmatically unless we decide to make the metadata public API. The nice thing is we can choose to do that whenever we want without breaking anyone. It doesn't even have to stay `RouteFilterMetadata`.

BrennanConroy commented on src/Http/Routing/src/EndpointDataSource.cs:
```diff
@@ -22,4 +24,149 @@ public abstract class EndpointDataSource
+    }
+
+    // We don't implement DebuggerDisplay directly on the EndpointDataSource base type because this could have side effects.
+    // REVIEW: Should we just make this public/protected? Derived types that can provide endpoints without side effects might find this useful.
```
I'd keep it internal for now, no need to add more to this change 😄

BrennanConroy commented on src/Http/Routing/src/Patterns/RoutePatternFactory.cs:
```diff
@@ -1186,6 +1152,42 @@ static Action<RoutePatternParameterPart> CheckDuplicateParameters(int parameterC
+            return leftList;
+        }
+
+        var leftCount = leftList.Count;
```
Is this saving something? If it is we should order it as so:
```
var leftCount = leftList.Count;
if (leftCount is 0)
{
    return rightList;
}
var rightCount = rightList.Count;
if (rightCount is 0)
{
    return leftList;
}
```

halter73 replied:
Not really. I was just referencing `leftList.Count` and `rightList.Count` a lot (3 times each) below so this looked a little cleaner to me. I'll adjust it to save the extra getter call. Technically, it could be slow given a bad implementation, but I doubt it is in practice.

BrennanConroy commented on src/Http/Routing/src/EndpointDataSource.cs:
```diff
@@ -22,4 +24,149 @@ public abstract class EndpointDataSource
+
+            // The RoutePattern, RequestDelegate, Order and DisplayName can all be overridden by non-group-aware conventions.
+            // Unlike with metadata, if a convention is applied to a group that changes any of these, I would expect these
+            // to be overridden as there's reasonable way to merge these properties.
```
```suggestion
            // to be overridden as there's no reasonable way to merge these properties.
```

BrennanConroy commented on src/Http/Routing/src/EndpointDataSource.cs:
```diff
@@ -22,4 +24,149 @@ public abstract class EndpointDataSource
+    /// <returns>
+    /// Returns a read-only collection of <see cref="RouteEndpoint"/> instances given the specified group <see cref="RouteGroupContext.Prefix"/> and <see cref="RouteGroupContext.Conventions"/>.
+    /// </returns>
+    public virtual IReadOnlyList<Endpoint> GetEndpointGroup(RouteGroupContext context)
```
Can this also be `RouteEndpoint` like `RouteEndpointDataSource` is doing?

halter73 replied:
It could, and it originally was, but that's just unnecessarily restrictive.

My reason for originally having it more restrictive is because, the default implementation will throw for any non `RouteEndpoint` anyway, and if you're going to prefix with a `RoutePattern`, don't you want the `Endpoint` type with a `RoutePattern`?

The reason I ultimately decided against it is because you can override the default implementation, and maybe you can meaningfully prefix a custom endpoint or at least apply conventions to it.

The bigger thing is no caller to `GetEndpointGroup` (I still prefer `GetGroupedEndpoints`, but I digress) actually needs the return type to only be `RouteEndpoint`s. That's the biggest reason I decided against it.

I kept `RouteEndpointDataSource` returning `IReadOnlyList<RoutePattern>` because its internal and was a chance to show off an obscure-ish covariant return feature introduced by C# 9, so it doesn't limit any public consumers. It has no practical benefit though.

davidfowl replied:
Why does the doc comment say it returns a RouteEndPoint, is that a typo?

davidfowl replied:
I also think we should rename this to what it was originally `GetGroupedEndpoints`

BrennanConroy commented on src/Http/Routing/src/RouteEndpointDataSource.cs:
```diff
@@ -0,0 +1,264 @@
+    {
+        if (_routeEntries.Count != 1)
+        {
+            throw new InvalidOperationException($"There are {_routeEntries.Count} endpoints defined! This can only be called for a single endpoint.");
```
I didn't see a test for this error case. Is it because it's not really possible today?

halter73 replied:
This really only called by our tests. I put this `throw` here only to make things easier on ourselves later if we mess up authoring a test and define multiple endpoints but expect a single builder..

There are a bunch of existing tests that define a single endpoint using `MapGet` or similar and want to get a single `RouteEndpointBuilder`. They all only define one endpoint, because defining more would just make it harder to get the right builder.

halter73 committed:
cleanup

halter73 committed:
polish

- less required :(

halter73 committed:
Clean up doc comments

davidfowl commented on src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:
```diff
@@ -5720,27 +5720,32 @@ public void Create_AddsDelegateMethodInfo_AsMetadata()
 
     [Fact]
-    public void Create_AddsDelegateAttributes_AsMetadata()
+    public void Create_DoesNotAddDelegateAttributes_AsMetadata()
```
These are interesting changes. So the caller of RDF.Create is now responsible for adding metadata, is that 6.0 behavior basically?

halter73 replied:
Yes. We were trying to make the RDF more "helpful" in .NET 7 (I'm pretty sure this was my idea), by adding more metadata for you.

I figured few if any callers other than `EndpointRouteBuilderExtensions` bothered adding the `MethodInfo` anyway. I'm glad we didn't go through with it obviously. Because the lack of flexibility meant that `WithOpenApi()` on a group couldn't read the route handler's `MethodInfo` metadata because it hadn't beed added by RDF yet.

davidfowl replied:
Yep!

davidfowl commented on src/Http/Routing/src/Builder/RouteHandlerFilterExtensions.cs:
```diff
@@ -56,27 +57,77 @@ public static RouteHandlerBuilder AddFilter(this RouteHandlerBuilder builder, IR
-        builder.RouteHandlerFilterFactories.Add(filterFactory);
+        builder.Add(endpointBuilder =>
+        {
+            // Even if we split RouteFilterMetadata across multiple metadata objects, RouteEndointDataSource will still use all of them.
+            // We're just trying to be good citizens and keep the Endpoint.Metadata as small as possible.
+            RouteHandlerFilterMetadata? filterMetadata = null;
+
+            foreach (var item in endpointBuilder.Metadata)
+            {
+                if (item is RouteHandlerFilterMetadata metadata)
+                {
+                    // Don't break early in case there is somehow higher priority metadata later in the list.
+                    filterMetadata = metadata;
+                }
+            }
+
+            if (filterMetadata is null)
+            {
+                filterMetadata = new();
+                endpointBuilder.Metadata.Add(filterMetadata);
+            }
+
+            filterMetadata.FilterFactories.Add(filterFactory);
```
I don't know that we do this anywhere else but I dunno if it matters much.

davidfowl commented on src/Http/Routing/src/CompositeEndpointDataSource.cs:
nit: typo

```suggestion
    private void CreateEndpointsUnsynchronized()
```

Also don't merge this suggestion because you need to change the call site too.

davidfowl commented on src/Http/Routing/src/EndpointDataSource.cs:
```diff
@@ -22,4 +24,148 @@ public abstract class EndpointDataSource
+
+            // Endpoint does not provide a RoutePattern but RouteEndpoint does. So it's impossible to apply a prefix for custom Endpoints.
+            // Supporting arbitrary Endpoints just to add group metadata would require changing the Endpoint type breaking any real scenario.
+            if (endpoint is not RouteEndpoint routeEndpoint)
+            {
+                throw new NotSupportedException(Resources.FormatMapGroup_CustomEndpointUnsupported(endpoint.GetType()));
+            }
```
Throw vs skip?

davidfowl commented on src/Http/Routing/src/EndpointDataSource.cs:
```diff
@@ -22,4 +24,148 @@ public abstract class EndpointDataSource
+    public virtual IReadOnlyList<Endpoint> GetEndpointGroup(RouteGroupContext context)
+    {
+        // Only evaluate Endpoints once per call.
+        var endpoints = Endpoints;
```
The dependency between `Endpoints` and `GetEndpointGroup` is a little scary. If you build a data source that implements Endpoints by calling `GetEndpointGroup`, it'll explode.

davidfowl commented on src/Http/Routing/src/RouteEndpointDataSource.cs:
```diff
@@ -0,0 +1,262 @@
+        }
+
+        RequestDelegate? factoryCreatedRequestDelegate = null;
+        RequestDelegate redirectedRequestDelegate = context =>
+        {
+            if (factoryCreatedRequestDelegate is null)
+            {
+                throw new InvalidOperationException(Resources.RouteEndpointDataSource_RequestDelegateCannotBeCalledBeforeBuild);
+            }
+
+            return factoryCreatedRequestDelegate(context);
+        };
```
What is this for? How would it show up?

davidfowl replied:
I see now. If a convention decorates the request delegate. Then we can't set overwrite the instance. The null check is just an extra safeguard but in practice nobody would call this before it's built.

davidfowl commented on src/Http/Routing/src/RouteEndpointDataSource.cs:
```diff
@@ -0,0 +1,262 @@
+        };
+
+        // We own EndpointBuilder.Metadata (in another assembly), so we know it's just a List.
+        var metadata = (List<object>)builder.Metadata;
```
Is this just to call AddRange?

davidfowl commented on src/Http/Routing/src/RouteEndpointDataSource.cs:
```diff
@@ -0,0 +1,262 @@
+        // Add MethodInfo as first metadata item
+        builder.Metadata.Add(handler.Method);
+
+        if (entry.HttpMethods is not null)
+        {
+            builder.Metadata.Add(new HttpMethodMetadata(entry.HttpMethods));
+        }
```
Why is this added before group conventions run?

davidfowl replied:
Found in the description:

> Even though RequestDelegateFactory now runs after any conventions are added to the endpoint (aside from conventions added by the RequestDelegateFactory), WithOpenApi is fixed by having the RouteEndointDataSource add MethodInfo before running any conventions. RouteEndointDataSource also adds any attributes as metadata. This is more similar to the previous behavior of RequestDelegateFactory in .NET 6 where it did not add this metadata.

Add this as a comment please 😄

davidfowl commented on src/Http/Routing/src/RouteEndpointDataSource.cs:
```diff
@@ -0,0 +1,262 @@
+            metadata.AddRange(attributes);
+        }
+
+        entry.Conventions.HasBeenBuilt = true;
```
I was confused by HasBeenBuilt. Where is the call to Build? Should this be `IsReadOnly` instead? Also add a comment about why this place in the code is where you can no longer add conventions (because we're going to be using them now and new ones won't be observed etc)

davidfowl commented on src/Http/Routing/src/RouteGroupBuilder.cs:
```diff
@@ -4,137 +4,116 @@
-        public override IChangeToken GetChangeToken() => new CompositeEndpointDataSource(_groupRouteBuilder._dataSources).GetChangeToken();
+        public override IChangeToken GetChangeToken()
+        {
+            if (_routeGroupBuilder._dataSources.Count is 0)
+            {
+                return NullChangeToken.Singleton;
+            }
+
+            if (_routeGroupBuilder._dataSources.Count is 1)
+            {
+                return _routeGroupBuilder._dataSources[0].GetChangeToken();
+            }
```
Switch expressions are much cleaner.

davidfowl commented on src/Http/Routing/src/RouteEndpointDataSource.cs:
```diff
@@ -0,0 +1,262 @@
+        return entry.Conventions;
+    }
+
+    public override IReadOnlyList<RouteEndpoint> Endpoints
```
I was so confused by this and only remember that C# allows for this now 😄

davidfowl commented on src/Http/Routing/src/EndpointRoutingMiddleware.cs:
```diff
@@ -26,6 +26,7 @@ internal sealed partial class EndpointRoutingMiddleware
         MatcherFactory matcherFactory,
         ILogger<EndpointRoutingMiddleware> logger,
         IEndpointRouteBuilder endpointRouteBuilder,
+        EndpointDataSource rootCompositeEndpointDataSource,
```
Can't we resolve this elsewhere? Doesnt it get resolved elsewhere today?

davidfowl commented on src/Http/Routing/src/RouteEndpointDataSource.cs:
```diff
@@ -0,0 +1,262 @@
+        // Otherwise, we always use the default of 0 unless a convention changes it later.
+        var order = entry.IsFallback ? int.MaxValue : 0;
+
+        RouteEndpointBuilder builder = new(redirectedRequestDelegate, pattern, order)
```
So thinking about this more I think I'd like to see one change to clean up how we're handling filters currently. Instead of treating it like metadata that we have to hunt from the collection, create a `RouteHandlerEndpointBuilder` that has the list of filters instead (we would unseal RouteEndpointBuilder).

```C#
public sealed class RouteHandlerEndpointBuilder : RouteEndpointBuilder
{
         public IReadOnlyList<Func<RouteHandlerContext, RouteHandlerFilterDelegate, RouteHandlerFilterDelegate>>? RouteHandlerFilterFactories { get; init; }
  
     public RouteHandlerEndpointBuilder(
       RequestDelegate requestDelegate,
       RoutePattern routePattern,
       int order) : base(requestDelegate, routePattern, order) { }
}
```

I like this for 2 reasons:
- It's a pattern for identifying these endpoints specifically so we can choose to skip applying metadata to endpoints based on this type.
- No weird merging of metadata entries is required.

halter73 replied:
That's a really good idea. We'd have to unseal `RouteEndpointBuilder`, but big deal. I think we can do one better though.

We could consider adding `RouteHandlerFilterFactories` directly to `RouteEndpointBuilder`. 

`RequestDelegate` is just a special case of our `Delegate` handling from a functional perspective. We should still keep things as they are today if there are no filters to avoid any possible startup penalty and trimming issues. But if someone does apply a filter to a `RequestDelegate` call to `MapPost(...)`, etc..., they should get to have their filter run on it if they want to.

I know this also means if someone calls something `MapControllers()`, `MapHub<Hub>` in a filtered group, the filters will run on those endpoints too, but I think most developers will actually really enjoy seeing the all registered routes and stuff in there filters if they filter an entire group. It's easy to separate Controllers and Hubs from other routes in the group so they don't get filtered. It's also easy enough for a filter to skip over route handlers `Delegate`s that are really just `RequestDelegate`s if they want to and that way they incur no per-request performance penalty.

This would involve calling RDF.Create() in `RouteEndpointBuilder.Build()` if and only if there are filters set. I know it might sound like going to far, but I really think it's super powerful and worth doing if we make it clear this is how filters work, and make it very easy to choose to no-op in your filter for plain `RequestDelegate`'s if you want to and incur no per request performance penalty. People writing their own debugging and diagnostic tools are going to love it. I might write a few now to show it off.

halter73 replied:
I've done this. See [this commit](https://github.com/dotnet/aspnetcore/pull/42195/commits/1aa70bb60e8d64b748c8fb8a45cbbf121b3ce2fb) for most of the implementation. Absolutely zero cost per-request unless a filter is added directly to the `RouteEndpointBuilder` *and* the filter actually modifies the per-invocation `RouteHandlerFilterDelegate`. If the filter does modify `RouteHandlerFilterDelegate` for the `RequestDelegate` endpoint, the modified `RouteHandlerFilterDelegate` does run of course, but that's what you asked for. A `RequestDelegate` is just one kind a `Delegate` after all. And aside from applying filters and handling `Task<T>`-returning `RequestDelegate`s (for more than just `MapGet` after that commit I might add), RDF does not change the behavior of the `RequestDelegate` compared to just running it directly.

This is all new API in .NET 7. No one is forcing anyone to `AddRouteHandlerFilter`s to everyday plain-`RequestDelegate` backed `RouteEndpointBuilder`s, but if you do, it's' very nice to have them run. What's really nice is this allows filters to run on practically any endpoint in existence. `MapHub`, `MapHealthChecks`, `MapControllers`, `MapRazorPages`, etc... But again, only if you ask for it. This is the first time people can even try doing this now that `AddRouteHandlerFilter` can be applied to any `IEndpointConventionBuilder`.

davidfowl replied:
I think this should be decoupled from this PR and the implementation shouldn't use RDF.

captainsafia replied:
Throwing in my two cents here now that I've reviewed the change. While I can recognize that there might be compelling scenarios for it, I'm a little cautious about:

- Changing our direction on whether or not route handler-filters could be applied to non-route handlers. We were committed enough to that stance to rename the API from `AddFilter` to `AddRouteHandlerFilter`. Side note: if we do want to pursue this direction then we should consider changing it back to `AddFilter`. 
- Invoking the RDF in all route endpoint builders. This change accounts invoking RDF from non-route handler scenarios when there are filters involved but have we considered other changes in the RDF that need to be made so that it can be invoked in more places.

In general, this seems high-risk/medium-reward and would be more sensible to approach as an independent change.

davidfowl commented on src/Http/Routing/src/RouteEndpointDataSource.cs:
```diff
@@ -0,0 +1,262 @@
+        }
+
+        // Let's see if any of the conventions added a filter before creating the RequestDelegate.
+        List<Func<RouteHandlerContext, RouteHandlerFilterDelegate, RouteHandlerFilterDelegate>>? routeHandlerFilterFactories = null;
+
+        foreach (var item in builder.Metadata)
+        {
+            if (item is RouteHandlerFilterMetadata filterMetadata)
+            {
+                routeHandlerFilterFactories ??= new();
+                routeHandlerFilterFactories.AddRange(filterMetadata.FilterFactories);
+            }
+        }
```
See my comment above about what I think would be a little cleaner than this.

halter73 committed:
Address straightforward PR feedback

halter73 committed:
Add filters to RouteEndpointBuilder

halter73 committed:
Merge remote-tracking branch 'origin/main' into halter73/41427

halter73 committed:
Remove redundant RequestDelegate returnType check

halter73 committed:
Fix merge conflict with RDF AOT changes

- also fix comment

halter73 committed:
Add missed MapHealthChecks suppression

davidfowl commented on src/Http/Http.Extensions/src/RequestDelegateFactoryOptions.cs:
```diff
@@ -51,4 +51,8 @@ public sealed class RequestDelegateFactoryOptions
     /// </remarks>
     public IList<object>? EndpointMetadata { get; init; }
+
+    // TODO: Add a RouteEndpointBuilder property and remove the EndpointMetadata property. Then do the same in RouteHandlerContext, EndpointMetadataContext
+    // and EndpointParameterMetadataContext. This will allow seeing the entire route pattern if the caller chooses to allow it.
+    // We'll probably want to add the RouteEndpointBuilder constructor without a RequestDelegate back and make it public too.
```
I'm not sure I like this coupling. RDF exists to be decoupled from routing.

davidfowl commented on src/Http/Routing/src/RouteEndpointBuilder.cs:
```diff
@@ -39,15 +44,39 @@ public sealed class RouteEndpointBuilder : EndpointBuilder
+        // ModelEndpointDataSource (Map(RoutePattern, RequestDelegate) and by extension MapHub, MapHealChecks, etc...),
+        // ActionEndpointDataSourceBase (MapControllers, MapRozorPages, etc...) and people with custom data sources or otherwise manually building endpoints
+        // using this type. At the moment this class is sealed, so at the moment we do not need to concern ourselves with what derived types may be doing.
+        if (RouteHandlerFilterFactories is { Count: > 0 })
+        {
+            // Even with filters applied, RDF.Create() will return back the exact same RequestDelegate instance we pass in if filters decide not to modify the
+            // invocation pipeline. We're just passing in a RequestDelegate so none of the fancy options pertaining to how the Delegate parameters are handled
+            // do not matter.
+            RequestDelegateFactoryOptions rdfOptions = new()
+            {
+                RouteHandlerFilterFactories = RouteHandlerFilterFactories,
+                EndpointMetadata = Metadata,
+            };
+
+            // We ignore the returned EndpointMetadata has been already populated since we passed in non-null EndpointMetadata.
+            requestDelegate = RequestDelegateFactory.Create(requestDelegate, rdfOptions).RequestDelegate;
+        }
```
While I think this is cool and I now understand it more, can we do this in a later change after we discuss the implications? If we enable this it might make sense to rename `RouteHandlerFilterInvocationContext` as it makes things more general purpose. I also am not sure about the coupling to the request delegate factory. That seems unnecessary. I would just rewrite the logic here since there's no need for codegen.

davidfowl commented on src/Http/Routing/src/EndpointDataSource.cs:
```diff
@@ -22,4 +24,148 @@ public abstract class EndpointDataSource
+            var fullRoutePattern = RoutePatternFactory.Combine(context.Prefix, routeEndpoint.RoutePattern);
+
+            // RequestDelegate can never be null on a RouteEndpoint. The nullability carries over from Endpoint.
+            var routeEndpointBuilder = new RouteEndpointBuilder(routeEndpoint.RequestDelegate!, fullRoutePattern, routeEndpoint.Order)
```
You need to copy the filters now as well right?

captainsafia commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -237,25 +252,38 @@ private static FactoryContext CreateFactoryContext(RequestDelegateFactoryOptions
+        if (filterPipeline is null && factoryContext.Handler is RequestDelegate)
+        {
+            // Make sure we're still not handling a return value.
+            if (!returnType.IsGenericType || returnType.GetGenericTypeDefinition() != typeof(Task<>))
```
Are you trying to validate that the return type wasn't modified by the logic above here? I'm not sure what value this check is adding.

davidfowl commented on src/Http/Routing/src/RouteEndpointDataSource.cs:
```diff
@@ -0,0 +1,257 @@
+    // For testing
+    internal RouteEndpointBuilder GetSingleRouteEndpointBuilder()
+    {
+        if (_routeEntries.Count is not 1)
```
This syntax.... @jaredpar have you seen this pattern much?
